src/main/java/core/basesyntax/MyHashMap.java

```diff
@@ -1,19 +1,113 @@
 package core.basesyntax;
 
 public class MyHashMap<K, V> implements MyMap<K, V> {
+    private static final int DEFAULT_INITIAL_CAPACITY = 16;
+    private static final float DEFAULT_LOAD_FACTOR = 0.75f;
+    private static final int DEFAULT_RESIZE_VALUE = 2;
+
+    private Node<K, V>[] table;
+    private int capacity;
+    private int size;
+    private int threshold;
+
+    static class Node<K, V> {
+        private final int hash;
+        private final K key;
+        private V value;
+        private Node<K, V> next;
+
+        Node(int hash, K key, V value, Node<K, V> next) {
+            this.hash = hash;
+            this.key = key;
+            this.value = value;
+            this.next = next;
+        }
+    }
 
     @Override
     public void put(K key, V value) {
-
+        if (table == null) {
+            initialize(DEFAULT_INITIAL_CAPACITY);
+        }
+        boolean keyPresent = updateIfKeyPresent(key, value);
+        if (!keyPresent) {
+            if (size >= threshold) {
+                resize();
+            }
+            createNewNode(key, value);
+        }
     }
 
     @Override
     public V getValue(K key) {
+        if (table != null) {
+            int bucketNumber = hash(key);
+            Node<K, V> nodeInBucket = table[bucketNumber];
+            while (nodeInBucket != null) {
+                if (keysAreEqual(nodeInBucket.key, key)) {
+                    return nodeInBucket.value;
+                }
+                nodeInBucket = nodeInBucket.next;
+            }
+        }
         return null;
     }
 
     @Override
     public int getSize() {
-        return 0;
+        return size;
+    }
+
+    private int hash(Object key) {
+        return key == null ? 0 : Math.abs(key.hashCode() % capacity);
+    }
+
+    private boolean updateIfKeyPresent(K key, V value) {
+        int hash = hash(key);
+        Node<K, V> nodeInBucket = table[hash];
+        while (nodeInBucket != null) {
+            if (keysAreEqual(key, nodeInBucket.key)) {
+                nodeInBucket.value = value;
+                return true;
+            }
+            nodeInBucket = nodeInBucket.next;
+        }
```

Comment:
It seems that updateIfKeyPresent and getValue have very similar logic, let's try to refactor this part to avoid code duplication 
![image](https://github.com/mate-academy/jv-hash-map/assets/46851188/6e9d61ca-9afd-439a-9a3b-a08a3d99d9a7)